File: Preparation-2023/two-sum-iv-input-is-a-bst/two-sum-iv-input-is-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    bool treetraverse(TreeNode* root,TreeNode* cur,int k){
        if(cur==NULL){
            return false;
        }
        
        bool a = searchtree(root,cur,k-cur->val);
        if(a){return a;}

        a = treetraverse(root,cur->left,k);
        if(a){return a;}
        a = treetraverse(root,cur->right,k);
        return a;
    }

    bool searchtree(TreeNode* root,TreeNode* cur, int k){
        if(root==NULL){return false;}
        if(root==cur){
            return false;
        }
        if(root->val==k){return true;}
        if(root->val > k){return searchtree(root->left,cur,k);}
        return searchtree(root->right,cur,k);
    }

    bool findTarget(TreeNode* root, int k) {
        return treetraverse(root,root,k);
    }

};

File: 653-two-sum-iv-input-is-a-bst/two-sum-iv-input-is-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        
        def inOrder(root):
            if root:
                yield from inOrder(root.left)
                yield root.val
                yield from inOrder(root.right)

        def inOrderReversed(root):
            if root:
                yield from inOrderReversed(root.right)
                yield root.val
                yield from inOrderReversed(root.left)

        leftGenerator = inOrder(root)
        rightGenerator = inOrderReversed(root)

        left, right = next(leftGenerator), next(rightGenerator)
        while left < right:
            if left + right == k: return True
            if left + right < k:
                left = next(leftGenerator)
            else:
                right = next(rightGenerator)
        return False
        